### PR TITLE
Add utils for core

### PR DESCRIPTION
**Pull request by isaac-tfn**
Describe Changes
---

Task Checklist
----
- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Have new dependencies been added?
  - [ ] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `pyproject.toml`

### CODE REVIEW

**Comment by isaac-tfn** [comment]
@wsp-mbuckley @BenTaylor-TfN I don't think I'll need to add any more to this for core, and until the next version if published all of the tests fail for core. Could one of you take a quick look if you get a chance?

**Comment by wsp-mbuckley** [comment]
> @wsp-mbuckley @BenTaylor-TfN I don't think I'll need to add any more to this for core, and until the next version if published all of the tests fail for core. Could one of you take a quick look if you get a chance?

Is it urgent or can it wait until Wednesday / Thursday?

**Comment by isaac-tfn** [comment]
> Is it urgent or can it wait until Wednesday / Thursday?

Wednesday or Thursday is fine I just think it should be this week so that core is as ready as possible to hand over next week.

**Comment by BenTaylor-TfN** [comment]
@wsp-mbuckley, can you cover the config_base files and I'll get the rest?

File: src/caf/toolkit/config_base.py [hunk 1]
```diff
@@ -220,6 +220,39 @@ def write_example(
...
+    """
+    if isinstance(object, list):
+        return True
+    elif isinstance(object, tuple):
+        return True
+    elif isinstance(object, set):
+        return True
+    elif isinstance(object, dict):
+        return True
+    return False
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/config_base.py:237]: Can take a tuple of types
```suggestion
return isinstance(object, (list, tuple, set, dict))
```

File: src/caf/toolkit/config_base.py [hunk 2]
```diff
@@ -220,6 +220,39 @@ def write_example(
 
 
 # # # FUNCTIONS # # #
+def _is_collection(object: any):
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/config_base.py:223]: Don't use `object` as a variable name - it's already the name of a built in.

Try `obj` instead?

File: src/caf/toolkit/config_base.py [hunk 3]
```diff
@@ -220,6 +220,39 @@ def write_example(
...
+
+def _remove_none_collection(data: Union[list, set, tuple]):
+    """Remove items recursively from collections which are None"""
+    # empty list, type set later
+    filtered = []
+    for item in data:
+        if item is None:
+            continue
+        elif isinstance(item, dict):
+            item = _remove_none_dict(item)
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/config_base.py:248]: Don't need the first `elif` if you're using `continue`

```suggestion
        if item is None:
            continue
           
        if isinstance(item, dict):
            item = _remove_none_dict(item)
```

File: src/caf/toolkit/translation.py [hunk 4]
```diff
@@ -853,6 +853,37 @@ def pandas_vector_zone_translation(
...
+    ----------
+    vector: The data you want to be translated. The zone system must form the index.
+    translation: Translation from caf.space. Data should be in columns (not index).
+    from_col: Column containing zone index of vector's zone system. Name must
+    match vector index name
+    to_col: Column containing zone index to be translated to.
+    factors_col: Column containning relevant factors (i.e. factors to translate
+    from 'from_col' to 'to_col'
+    """
+    translation.set_index([from_col, to_col], inplace=True)
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/translation.py:877]: Do we want to do this inplace? It means changing an object that was given to this class.

File: src/caf/toolkit/translation.py [hunk 5]
```diff
@@ -853,6 +853,37 @@ def pandas_vector_zone_translation(
...
 
 
+def pandas_multi_column_translation(
+    vector: pd.DataFrame,
+    translation: pd.DataFrame,
+    from_col: str,
+    to_col: str,
+    factors_col: str,
+):
+    """
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/translation.py:863]: I feel like this function is too specific to CAF.core to belong here.
I'd say either it should be generalised in-line with other functions in this module, or moved into core. (I'd rather it be generalised tbh)

I assumed it took a DF with the zoning system as the index and translated all other columns. Some questions:
- How do we know the df index is in the same order for the translation and the given vector?
- How is this different to just calling the vector translation on each column of `vector`?
  - Is it slower than the above method as you're using pandas methods rather than pure numpy?

File: src/caf/toolkit/translation.py [hunk 6]
```diff
@@ -853,6 +853,37 @@ def pandas_vector_zone_translation(
...
+    Translate a multicolumn dataframe between zone systems using a caf.space
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/translation.py:864]: Initial docstring text should be in line with the opening `"""` and take up no more than 80 characters.

**Review by BenTaylor-TfN — CHANGES_REQUESTED** [review]

File: src/caf/toolkit/translation.py (hunk 4 above)
Comment (isaac-tfn) [reply src/caf/toolkit/translation.py:877]: > Do we want to do this inplace? It means changing an object that was given to this class.

No reason to no, this function was very much written for the current use case and for caf.core it doesn't matter

**Comment by wsp-mbuckley** [comment]
Ben's comments cover everything in the `config_base` module, I would just add it would be worth adding a test to make sure the converting None values inside dicts or collections works

**Comment by BenTaylor-TfN** [comment]
I've refactored this code a little bit to make it more flexible with multiple frontends.
I've got a few more comments to add. @isaac-tfn, please can you address the requested changes when you're back.

File: src/caf/toolkit/translation.py [hunk 7]
```diff
@@ -853,6 +1034,131 @@ def pandas_vector_zone_translation(
...
+    vector: pd.DataFrame,
+    translation: pd.DataFrame,
+    translation_from_col: str,
+    translation_to_col: str,
+    translation_factors_col: str,
+    from_unique_index: list[Any] = [],
+    to_unique_index: list[Any] = [],
+    check_totals: bool = True,
+) -> pd.DataFrame:
+    """Efficiently translate a multi-column pandas vector between index systems.
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/translation.py:1047]: Please check this docstring and update accordingly to ensure its accurate. I've just copied and pasted it from the single-vector for now

File: src/caf/toolkit/translation.py [hunk 8]
```diff
@@ -853,6 +1034,131 @@ def pandas_vector_zone_translation(
...
 
 
+def pandas_multi_vector_zone_translation(
+    vector: pd.DataFrame,
+    translation: pd.DataFrame,
+    translation_from_col: str,
+    translation_to_col: str,
+    translation_factors_col: str,
+    from_unique_index: list[Any] = [],
+    to_unique_index: list[Any] = [],
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/translation.py:1044]: Never use mutable types (such as lists) as default arguments! Use None instead, its much safer.

**Review by BenTaylor-TfN — CHANGES_REQUESTED** [review]

**Comment by BenTaylor-TfN** [comment]
Rebased on main.

Will update code now to reflect discussion in #79.

File: src/caf/toolkit/translation.py [hunk 9]
```diff
@@ -707,6 +767,42 @@ def pandas_matrix_zone_translation(
...
+    translation_factors_col: str,
+    from_unique_index: list[Any],
+    to_unique_index: list[Any],
+    translation_dtype: Optional[np.dtype] = None,
+    vector_infill: float = 0.0,
+    translate_infill: float = 0.0,
+    check_totals: bool = True,
+) -> pd.DataFrame:
+    # pylint: disable=too-many-arguments
+    ...  # pragma: no cover
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/translation.py:803]: This second overload is currently causing the following error with MyPy linting:
image

This happens because we don't have the pandas-stubs installed when performing mypy checks, meaning all pandas type-hints are assumed to be `Any`. 

Best solution is to install pandas-stubs to remove this issue. However, this introduces a number of other errors we have overlooked until now. I've open a new PR (#78) that deals with this.
